src/main.rs

```diff
@@ -47,55 +51,83 @@ fn compile(txt: &str) -> (String, Vec<usize>) {
     (out, map)
 }
 
-fn interp(prg: &[u8], map: Vec<usize>) {
-    let mut pc = 0;
-    let mut ptr = 0;
-    let mut cells = Vec::with_capacity(30000);
-    cells.resize(30000, 0u8);
-    let mut stdin = std::io::stdin().bytes();
-    let mut stdout = std::io::stdout();
-    while pc < prg.len() {
-        match prg[pc] as char {
-            '>' => {
-                if ptr == cells.len() {
-                    cells.push(0);
-                }
-                ptr += 1;
-            }
-            '<' => {
-                if ptr > 0 {
-                    ptr -= 1;
-                }
-            }
-            '+' => {
-                cells[ptr] = cells[ptr].wrapping_add(1);
-            }
-            '-' => {
-                cells[ptr] = cells[ptr].wrapping_sub(1);
+#[derive(Debug)]
+struct InterpCx {
+    pc: usize,
+    ptr: usize,
+    cells: Vec<u8>,
+    prog: Vec<u8>,
+    map: Vec<usize>,
+    stdin: Bytes<Stdin>,
+    stdout: Stdout,
+}
+
+fn interp(prog: &[u8], map: Vec<usize>) {
+    let mut icx = InterpCx {
+        pc: 0,
+        ptr: 0,
+        cells: Vec::from([0; 30000]),
+        map,
+        prog: prog.to_vec(),
+        stdin: std::io::stdin().bytes(),
+        stdout: std::io::stdout(),
+    };
+
+    // FIXME only use locations for loop heads.
+    let mut locs = Vec::with_capacity(icx.prog.len());
```

Comment:
Is this `let mut locs = vec![Location::new(); icx.prog.len()]` in disguise? I can never quite remember the semantics of vec initialisation!

Reply:
`icx.prog.len()` isn't constant.

This has changed in 4137a7d8 anyway.

Reply:
> icx.prog.len() isn't constant.

I'm not sure if `vec!` can deal with that or not, but the fact that I can't remember does suggest that using `vec!` in such a way wouldn't be a good idea anyway!

Reply:
I just investigated.

```
let _: Vec<_> = vec![(); env::args().len()];
```

^ works.

```
Vec::from([(); env::args().len()]);
```

^ does not

Reply:
Lesson learned -- thanks :)